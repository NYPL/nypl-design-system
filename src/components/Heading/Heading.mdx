import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as HeadingStories from "./Heading.stories";
import Link from "../Link/Link";

<Meta of={HeadingStories} />

# Heading

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.0.4`    |
| Latest            | `1.7.0`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#updated-typographic-styles" target="_self">Updated Typographic Styles</Link>}
- {<Link href="#native-heading-elements" target="_self">Native Heading Elements</Link>}
- {<Link href="#size-styles" target="_self">Size Styles</Link>}
- {<Link href="#overline-and-subtitle" target="_self">Overline and Subtitle Elements</Link>}
- {<Link href="#heading-with-bold-text" target="_self">Heading with Bold Text</Link>}
- {<Link href="#heading-with-links" target="_self">Heading with Links</Link>}

## Overview

<Description of={HeadingStories} />

A `Heading` component will render a standard HTML `<h>` tag (1-6). The heading's
text can be passed in through a `text` prop or as a child. Default styles for
semantic elements can be overwritten using the `size` prop.

## Component Props

<Canvas of={HeadingStories.WithControls} />

<ArgTypes of={HeadingStories.WithControls} />

## Accessibility

When adding headings to a webpage, it is important to ensure that the heading
hierarchy is consistent. This means they should start with only one `h1` for the
page title and then proceed with `h2`s, `h3`s, `h4`s, `h5`s and `h6`s in the
proper order and not skipping any. For example, the following is invalid HTML:

<Source
  code={`
<h1>Page Title</h1>
<h2>Subtitle</h2>
<h4>Sub-subtitle</h4>
`}
  language="html"
/>

When either the overline or subtitle elements are rendered as part of the
heading, then the whole lockup will be wrapped in an `<hgroup>` element with
proper `role` and `aria-roledescription` attributes applied.

<Source
  code={`
<hgroup role="group" aria-roledescription="Heading group">
  <h2>Heading Title</h2>
  <p aria-roledescription="subtitle">Subtitle</p>
</hgroup>
`}
  language="html"
/>

Resources:

- [W3C WAI Headings](https://www.w3.org/WAI/tutorials/page-structure/headings/)
- [A11y Project Accessible heading structure](https://www.a11yproject.com/posts/how-to-accessible-heading-structure/)
- [Chakra UI Heading](https://chakra-ui.com/docs/components/typography/heading)

## Updated Typographic Styles

New typographic styles were introduced in June 2023 and the new styles are the
supported options. The old typographic styles are still available, but they have
been deprecated.

## Native Heading Elements

The `level` prop can be used to set a native heading element (ex. `<h1>`,
`<h2>`, ...).

<Source
  code={`
<Heading level="h1" />
`}
  language="jsx"
/>

The default text styles for the native heading element levels correspond
directly with the values of the `size` prop. For example, the `heading1` size is
the default style for the `<h1>` element. The `heading2` size is the default
style for the `<h2>` element. And so on.

Based on the June 2023 updates, the following values should be used for the
`level` prop.

`h1`, `h2`, `h3`, `h4`, `h5`, `h6`

<Canvas of={HeadingStories.DefaultStyles} />

### Deprecated Options

The older `level` values and the associated styles are still available, but the
values have been deprecated and they should only used to fulfill older design
requirements.

For the deprecated options, the `primary` size is the default style for the
`<h1>` element. The `secondary` size is the default style for the `<h2>`
element. The `tertiary` size is the default style for the `<h3>` element. And
the `callout` size is the default style for the `<h4>` element. `<h5>` and
`<h6>` do not have corresponding `size` styles.

`one`, `two`, `three`, `four`, `five`, `six`

<Canvas of={HeadingStories.DefaultStylesDeprecated} />

### Migrating from Old to New

When adopting the new `level` values, there is a direct relationship between the
deprecated values and the new values. For example, from a DOM persepctive, `two`
should be replaced with `h2`, as they will both render the `<h2>` element.
Please note that migrating to the new `level` values will change the styles for
the headings. Make sure to use the `size` prop to make adjustments as needed.

## Size Styles

The `size` prop can be used to override the default styles of the native heading
elements.

<Source
  code={`
<Heading size="heading1" />
`}
  language="jsx"
/>

Based on the June 2023 updates, the following values should be used for the
`size` prop.

`display1`, `heading1`, `heading2`, `heading3`, `heading4`, `heading5`,
`heading6`

For usage recommendations of each of the `size` options, please refer to the
[Typography Style
Guide](../?path=/docs/style-guide-typography--docs#size-options).

**Note:** In the examples below, the `level` prop is set to `level="h1"` for
each `Heading` component and the `size` prop is used to adjust the text size.

<Canvas of={HeadingStories.SizeStyles} />

### Deprecated Options

The older `size` values and the associated styles are still available, but the
values have been deprecated and they should only used to fulfill older design
requirements.

`primary`, `secondary`, `tertiary`, `callout`

**Note:** In the examples below, the `level` prop is set to `level="one"` for
each `Heading` component and the `size` prop is used to adjust the text size.

<Canvas of={HeadingStories.SizeStylesDeprecated} />

## Overline and Subtitle

The `Heading` component offers two additional text elements as part of the full
heading lockup: overline and subtitle. When the `overline` or `subtitle` props
are passed, the `Heading` component will handle the DOM structure and styling
for the corresponding text elements.

The styling of the overline and subtitle elements include differentiating the
font size for both elements based on the `size` prop. For `display1`,
`heading1`, and `heading2`, the text for these elements will be one size. And
for `heading3`, `heading4`, `heading5`, and `heading6`, the text for these
elements will be another (smaller) size.

**Note:** In the examples below, the `level` prop is set to `level="h1"` for
each `Heading` component and the `size` prop is used to adjust the text size.

### Overline

Use the `overline` prop to set the text for an eyebrow text element that sits
above the heading text. The overline text is styled to be all uppercase.

<Source
  code={`
<Heading
  overline="Overline"
  text="Lorem ipsum dolor"
/>
`}
  language="jsx"
/>

<Canvas of={HeadingStories.Overline} />

### Subtitle

Use the `subtitle` prop to set the text for a text element that sits below the
heading text.

<Source
  code={`
<Heading
  subtitle="The subtitle text sits below the main heading text."
  text="Lorem ipsum dolor"
/>
`}
  language="jsx"
/>

<Canvas of={HeadingStories.Subtitle} />

### Full Heading Lockup

Or use both the `overline` and `subtitle` props in combination to render a fully
composed heading lockup.

<Source
  code={`
<Heading
  overline="Overline"
  subtitle="The subtitle text sits below the main heading text."
  text="Lorem ipsum dolor"
/>
`}
  language="jsx"
/>

<Canvas of={HeadingStories.OverlineAndSubtitle} />

## Heading with Bold Text

<Canvas of={HeadingStories.BoldText} />

## Heading with Links

When the `url` prop is passed to `Heading`, a child `<a>` element is created and
the heading text becomes an active link.

<Canvas of={HeadingStories.Links} />
