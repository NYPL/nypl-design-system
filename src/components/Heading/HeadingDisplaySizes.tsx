export enum DisplaySizes {
  Primary = "primary",
  Secondary = "secondary",
  Tertiary = "tertiary",
  Callout = "callout",
}
