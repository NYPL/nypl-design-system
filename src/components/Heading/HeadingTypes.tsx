export enum HeadingSizes {
  Primary = "primary",
  Secondary = "secondary",
  Tertiary = "tertiary",
  Callout = "callout",
}

export enum HeadingLevels {
  One = "one",
  Two = "two",
  Three = "three",
  Four = "four",
  Five = "five",
  Six = "six",
}
