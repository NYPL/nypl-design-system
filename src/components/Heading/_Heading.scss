//
// Generic header styles:
// All arguments are optional. If not defined, the defaults below will be used
//

@mixin heading-xl($color: inherit) {
  color: #{$color};
  font-size: var(--font-size-4);
  font-weight: 300;
  letter-spacing: 0;
  line-height: 1.1;
  margin: 0 0 0.5em;
  width: auto;

  &__link {
    color: inherit;

    &:hover {
      color: inherit;
    }
  }
}

@mixin heading-large($color: inherit) {
  color: #{$color};
  font-size: var(--font-size-3);
  font-weight: 500;
  line-height: 1.25;
  margin: 0 0 0.5em;
  width: auto;

  &__link {
    color: inherit;

    &:hover {
      color: inherit;
    }
  }
}

@mixin heading-medium($color: inherit) {
  color: #{$color};
  font-size: var(--font-size-2);
  font-weight: 500;
  line-height: 1.25;
  margin: 0 0 0.5em;
  width: auto;

  &__link {
    color: inherit;

    &:hover {
      color: inherit;
    }
  }
}

@mixin heading-small($color: inherit) {
  color: #{$color};
  font-size: var(--font-size-1);
  font-weight: 500;
  line-height: 1.25;
  margin: 0 0 0.5em;
  width: auto;

  &__link {
    color: inherit;

    &:hover {
      color: inherit;
    }
  }
}

@mixin heading-xs($color: inherit) {
  color: #{$color};
  font-size: var(--font-size-1);
  font-weight: 400;
  line-height: 1.15;
  margin: 0 0 0.5em;
  width: auto;

  &__link {
    color: inherit;

    &:hover {
      color: inherit;
    }
  }
}

.nypl-ds {
  h1 {
    @include heading-xl;
  }

  h2 {
    @include heading-large;
  }

  h3 {
    @include heading-medium;
  }

  h4 {
    @include heading-small;
  }

  h5,
  h6 {
    @include heading-xs;
  }

  .heading {
    &--primary {
      @include heading-xl;
    }

    &--secondary {
      @include heading-large;
    }

    &--tertiary {
      @include heading-medium;
    }

    &--callout {
      @include heading-xs;
    }
  }
}

.heading__link {
  color: inherit;
  text-decoration: none;

  &:hover {
    text-decoration: underline;
  }
}

.heading-section {
  &__more-link {
    @include more-link;
  }

  &__heading {
    &--has-link {
      display: inline;
    }

    margin-right: var(--space-xs);
  }
}

.page-title {
  &__heading {
    &--block-color {
      background-color: var(--section-research-primary);
      color: var(--ui-white);
    }
  }
}
