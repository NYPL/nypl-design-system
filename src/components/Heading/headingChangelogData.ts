/** This data is used to populate the ComponentChangelogTable component.
 *
 * date: string (when adding new entry during development, set value as "Prerelease")
 * version: string (when adding new entry during development, set value as "Prerelease")
 * type: "Bug Fix" | "New Feature" | "Update";
 * affects: array["Accessibility" | "Documentation" | "Functionality" | "Styles"];
 * notes: array (will render as a bulleted list, add one array element for each list element)
 */
import { ChangelogData } from "../../utils/ComponentChangelogTable";

export const changelogData: ChangelogData[] = [
  {
    date: "2024-03-14",
    version: "3.0.0",
    type: "Update",
    affects: ["Styles"],
    notes: [
      "Chakra 2.8 update.",
      "Updates the component's theme file to use native Chakra responsive styles of the font sizes and removed the conditional logic that handled that previously.",
    ],
  },
  {
    date: "2024-02-22",
    version: "2.1.6",
    type: "Bug Fix",
    affects: ["Styles"],
    notes: ['Fixed the desktop font size for the "heading5" variant.'],
  },
  {
    date: "2024-02-05",
    version: "2.1.5",
    type: "Update",
    affects: ["Styles"],
    notes: [
      "Updated to use a custom `@media query` method to handle the responsive `font-size` styles.",
    ],
  },
  {
    date: "2023-12-07",
    version: "2.1.3",
    type: "Update",
    affects: ["Styles"],
    notes: [
      'Updates the `font-weight` to "regular" for the `subtitle1` and `subtitle2` text styles.',
    ],
  },
  {
    date: "2023-10-26",
    version: "2.1.1",
    type: "Update",
    affects: ["Accessibility"],
    notes: [
      'Updated the `aria-roledescription` value to "subtitle" (a more familiar and recognizable term) for the `overline` element.',
    ],
  },
  {
    date: "2023-9-28",
    version: "2.0.0",
    type: "Update",
    affects: ["Styles"],
    notes: ["Applied Typo2023 styles, including font size and font color."],
  },
];
