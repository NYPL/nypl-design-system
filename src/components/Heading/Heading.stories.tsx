import { VStack } from "@chakra-ui/react";
import type { Meta, StoryObj } from "@storybook/react";

import Heading, { headingLevelsArray, headingSizesArray } from "./Heading";
import Link from "../Link/Link";
import { argsBooleanType } from "../../helpers/storybookUtils";

const meta: Meta<typeof Heading> = {
  title: "Components/Typography & Styles/Heading",
  component: Heading,
  argTypes: {
    className: { control: false },
    id: { control: false },
    isCapitalized: argsBooleanType(),
    isUppercase: argsBooleanType(),
    isLowercase: argsBooleanType(),
    level: {
      control: { type: "select" },
      options: headingLevelsArray,
      table: { defaultValue: { summary: "h2" } },
    },
    noSpace: { table: { defaultValue: { summary: false } } },
    overline: { control: { type: "text" } },
    size: {
      control: { type: "select" },
      options: headingSizesArray,
    },
    subtitle: { control: { type: "text" } },
    text: { control: { type: "text" } },
    url: { control: { type: "text" } },
    urlClass: { control: false },
  },
};

export default meta;
type Story = StoryObj<typeof Heading>;

/**
 * Main Story for the Heading component. This must contains the `args`
 * and `parameters` properties in this object.
 */
export const WithControls: Story = {
  args: {
    className: undefined,
    id: "heading-id",
    isCapitalized: false,
    isUppercase: false,
    isLowercase: false,
    level: "h2",
    noSpace: true,
    overline: undefined,
    size: undefined,
    subtitle: undefined,
    text: "Default Heading",
    url: undefined,
    urlClass: undefined,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=10975%3A16",
    },
    jest: ["Heading.test.tsx"],
  },
  render: (args) => (
    <VStack align="left" spacing="l">
      <Heading {...args} />
      <Heading {...args} color="brand.primary" />
      <Heading
        {...args}
        bgColor="brand.primary"
        color="ui.typography.inverse.heading"
        px="s"
        py="xs"
      />
    </VStack>
  ),
};

// The following are additional Heading example Stories.
export const DefaultStyles: Story = {
  render: () => (
    <VStack align="left" spacing="l">
      <Heading
        id="heading-h1"
        level="h1"
        noSpace
        text="(h1) Lorem ipsum dolor"
      />
      <Heading id="heading-h2" level="h2" text="(h2) Lorem ipsum dolor" />
      <Heading id="heading-h3" level="h3" text="(h3) Lorem ipsum dolor" />
      <Heading id="heading-h4" level="h4" text="(h4) Lorem ipsum dolor" />
      <Heading id="heading-h5" level="h5" text="(h5) Lorem ipsum dolor" />
      <Heading id="heading-h6" level="h6" text="(h6) Lorem ipsum dolor" />
    </VStack>
  ),
};

export const SizeStyles: Story = {
  name: "Sizes",
  render: () => (
    <VStack align="left" spacing="l">
      <Heading
        id="heading-display1"
        level="h1"
        noSpace
        size="display1"
        text="(display1) Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading1"
        level="h1"
        size="heading1"
        text="(heading1) Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading2"
        level="h1"
        size="heading2"
        text="(heading2) Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading3"
        level="h1"
        size="heading3"
        text="(heading3) Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading4"
        level="h1"
        size="heading4"
        text="(heading4) Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading5"
        level="h1"
        size="heading5"
        text="(heading5) Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading6"
        level="h1"
        size="heading6"
        text="(heading6) Lorem ipsum dolor"
      />
    </VStack>
  ),
};

export const DefaultStylesDeprecated: Story = {
  name: "Deprecated Styles",
  render: () => (
    <VStack align="left" spacing="l">
      <Heading
        id="heading-one"
        level="one"
        noSpace
        text="(one) Lorem ipsum dolor"
      />
      <Heading id="heading-two" level="two" text="(two) Lorem ipsum dolor" />
      <Heading
        id="heading-three"
        level="three"
        text="(three) Lorem ipsum dolor"
      />
      <Heading id="heading-four" level="four" text="(four) Lorem ipsum dolor" />
      <Heading id="heading-five" level="five" text="(five) Lorem ipsum dolor" />
      <Heading
        id="heading-six"
        color="inheret"
        level="six"
        text="(six) Lorem ipsum dolor"
      />
    </VStack>
  ),
};

export const SizeStylesDeprecated: Story = {
  name: "Deprecated Sizes",
  render: () => (
    <VStack align="left" spacing="l">
      <Heading
        id="heading-primary"
        level="one"
        noSpace
        size="primary"
        text="(primary) Lorem ipsum dolor"
      />
      <Heading
        id="heading-secondary"
        level="one"
        size="secondary"
        text="(secondary) Lorem ipsum dolor"
      />
      <Heading
        id="heading-tertiary"
        level="one"
        size="tertiary"
        text="(tertiary) Lorem ipsum dolor"
      />
      <Heading
        id="heading-callout"
        level="one"
        size="callout"
        text="(callout) Lorem ipsum dolor"
      />
    </VStack>
  ),
};

export const Overline: Story = {
  render: () => (
    <VStack align="left" spacing="l">
      <Heading
        id="heading-display1"
        level="h1"
        noSpace
        overline="Overline"
        size="display1"
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading1"
        level="h1"
        noSpace
        overline="Overline"
        size="heading1"
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading2"
        level="h1"
        noSpace
        overline="Overline"
        size="heading2"
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading3"
        level="h1"
        noSpace
        overline="Overline"
        size="heading3"
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading4"
        level="h1"
        noSpace
        overline="Overline"
        size="heading4"
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading5"
        level="h1"
        noSpace
        overline="Overline"
        size="heading5"
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading6"
        level="h1"
        noSpace
        overline="Overline"
        size="heading6"
        text="Lorem ipsum dolor"
      />
    </VStack>
  ),
};

export const Subtitle: Story = {
  render: () => (
    <VStack align="left" spacing="l">
      <Heading
        id="heading-display1"
        level="h1"
        noSpace
        size="display1"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading1"
        level="h1"
        noSpace
        size="heading1"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading2"
        level="h1"
        noSpace
        size="heading2"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading3"
        level="h1"
        noSpace
        size="heading3"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading4"
        level="h1"
        noSpace
        size="heading4"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading5"
        level="h1"
        noSpace
        size="heading5"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading6"
        level="h1"
        noSpace
        size="heading6"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
    </VStack>
  ),
};

export const OverlineAndSubtitle: Story = {
  name: "Overline and Subtitle",
  render: () => (
    <VStack align="left" spacing="l">
      <Heading
        id="heading-display1"
        level="h1"
        overline="Overline"
        size="display1"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading1"
        level="h1"
        overline="Overline"
        size="heading1"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading2"
        level="h1"
        overline="Overline"
        size="heading2"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading3"
        level="h1"
        overline="Overline"
        size="heading3"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading4"
        level="h1"
        overline="Overline"
        size="heading4"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading5"
        level="h1"
        overline="Overline"
        size="heading5"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
      <Heading
        id="heading-heading6"
        level="h1"
        overline="Overline"
        size="heading6"
        subtitle="The subtitle text sits below the main heading text."
        text="Lorem ipsum dolor"
      />
    </VStack>
  ),
};

export const BoldText: Story = {
  render: () => (
    <Heading id="headingWithBold" level="h2" noSpace>
      <>
        Heading with a <b>Bold</b> Word
      </>
    </Heading>
  ),
};

export const Links: Story = {
  name: "Links",
  render: () => (
    <VStack align="left" spacing="l">
      <Heading
        id="heading-with-text-url-values"
        level="h3"
        noSpace
        text="Heading with `url` and `text` props"
        url="http://apple.com"
        urlClass="special-link"
      />
      <Heading
        id="heading-with-link-url"
        level="h3"
        noSpace
        url="http://apple.com"
        urlClass="special-link"
      >
        Heading with `url` prop value
      </Heading>
      <Heading id="heading-with-anchor" level="h3" noSpace>
        <a href="#hello">Link from HTML anchor element</a>
      </Heading>
      <Heading id="heading-with-DS-link" level="h3" noSpace>
        <Link href="#hello">Link using DS `Link` component</Link>
      </Heading>
      <Heading id="heading-with-link-child" level="h3" noSpace>
        <>
          Not acceptable. Do not use. <a href="#hello">Link</a>
        </>
      </Heading>
      <Heading id="heading-with-DS-link-child" level="h3" noSpace>
        <>
          <span>Text wrapped in a span </span>
          <Link href="#hello">Link</Link>
          <span> Not a recommended pattern</span>
        </>
      </Heading>
    </VStack>
  ),
};
