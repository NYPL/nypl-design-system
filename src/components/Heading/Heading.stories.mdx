import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Preview,
  Description,
} from "@storybook/addon-docs/blocks";
import { withDesign } from "storybook-addon-designs";

import Heading from "./Heading";
import { DisplaySizes } from "./HeadingDisplaySizes";

<Meta
  title="Heading"
  component={Heading}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url:
        "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=10975%3A16",
    },
  }}
  argTypes={{
    blockName: { table: { disable: true } },
    className: { table: { disable: true } },
    modifiers: { table: { disable: true } },
  }}
/>

# Heading

<Description of={Heading} />

A `Heading` component will render a standard HTML `<h>` tag (1-6). The heading's text can be passed in through a `text` prop or as a child. Default styles for semantic elements can be overwritten using the `displaySize` prop.

<Preview withToolbar>
  <Story
    name="Heading Props"
    args={{
      id: "heading1",
      level: 1,
      text: "Heading  1",
    }}
  >
    {args => <Heading {...args} />}
  </Story>
</Preview>

<ArgsTable story="Heading Props" />

## Default Heading Styles

<Preview>
  <Story name="Default Heading Styles">
    <div>
      <Heading id="heading1" level={1} text="<h1> Heading 1" />
      <Heading id="heading2" level={2} text="<h2> Heading 2" />
      <Heading id="heading3" level={3} text="<h3> Heading 3" />
      <Heading id="heading4" level={4} text="<h4> Heading 4" />
      <Heading id="heading5" level={5} text="<h5> Heading 5" />
      <Heading id="heading6" level={6} text="<h6> Heading 6" />
    </div>
  </Story>
</Preview>

## DisplaySize Styles

<Preview>
  <Story name="DisplaySize Styles">
    <div>
      <Heading
        id="heading1"
        level={1}
        displaySize={DisplaySizes.Primary}
        text="<h1> DisplaySizes.Primary"
      />
      <Heading
        id="heading1"
        level={1}
        displaySize={DisplaySizes.Secondary}
        text="<h1> DisplaySizes.Secondary"
      />
      <Heading
        id="heading1"
        level={1}
        displaySize={DisplaySizes.Tertiary}
        text="<h1> DisplaySizes.Tertiary"
      />
      <Heading
        id="heading1"
        level={1}
        displaySize={DisplaySizes.Callout}
        text="<h1> DisplaySizes.Callout"
      />
    </div>
  </Story>
</Preview>

## Heading with Bold

<Preview>
  <Story name="Heading with Bold">
    <Heading id="headingWithBold" level={2}>
      <>
        Heading with a <b>Bold</b> Word
      </>
    </Heading>
  </Story>
</Preview>

## Headings with Links

When the `url` prop is passed to `Heading`, a child `<a>` element is created and the heading text becomes an active link.

<Preview>
  <Story name="Headings with Links">
    <Heading
      id="headingWithLink"
      level={2}
      text="Heading with URL Props"
      url="http://apple.com"
      urlClass="special-link"
    />
    <Heading id="headingWithLink" level={2}>
      <>
        Heading with a Word <a href="hello">Link</a>
      </>
    </Heading>
  </Story>
</Preview>
