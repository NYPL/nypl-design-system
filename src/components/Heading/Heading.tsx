// MT-82, MT 225, etc
import * as React from "react";
import bem from "../../utils/bem";
import { DisplaySizes } from "./HeadingDisplaySizes";

export interface HeadingProps {
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** Optional className that appears in addition to `heading` */
  className?: string;
  /** Optional size used to override the default styles of the semantic HTML `<h>` elements */
  displaySize?: DisplaySizes;
  /** Optional ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Number 1-6; used to create the `<h*>` tag */
  level: number;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** Inner text of the `<h*>` element */
  text?: string;
  /** Optional URL that header points to; when `url` prop is passed to `Heading`, a child `<a>` element is created and the heading text becomes an active link */
  url?: string;
  /** Optional className for the URL when the `url` prop is passed */
  urlClass?: string;
}

export default function Heading(props: React.PropsWithChildren<HeadingProps>) {
  const {
    blockName,
    className,
    displaySize,
    id,
    level,
    modifiers = [],
    text,
    url,
    urlClass,
  } = props;

  const baseClass = "heading";

  if (displaySize) {
    modifiers.push(displaySize);
  }

  if (level < 1 || level > 6) {
    throw new Error("Heading only supports levels 1-6");
  }

  if (!props.children && !text) {
    throw new Error("Heading has no children, please pass prop: text");
  }

  if (React.Children.count(props.children) > 1) {
    const children = React.Children.map(
      props.children,
      child => (child as JSX.Element).type
    );
    // Catching the error because React's error isn't as helpful.
    throw new Error(
      `Please only pass one child into Heading, got ${children.join(", ")}`
    );
  }

  let content: string | React.ReactNode;
  if (props.children) {
    content = url
      ? React.createElement(
          "a",
          { href: url, className: urlClass },
          props.children
        )
      : props.children;
  } else {
    content = url
      ? React.createElement("a", { href: url, className: urlClass }, text)
      : text;
  }

  const headingProps = {
    className: bem(baseClass, modifiers, blockName, [className]),
    id: id,
  };

  return React.createElement("h" + level, headingProps, content);
}
