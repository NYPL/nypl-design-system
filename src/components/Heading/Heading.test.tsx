import { expect } from "chai";
import * as Enzyme from "enzyme";
import * as React from "react";

import Heading from "./Heading";
import { DisplaySizes } from "./HeadingDisplaySizes";

describe("Section Headings", () => {
  let wrapper: Enzyme.ShallowWrapper<any, any>;

  it("Shows heading", () => {
    wrapper = Enzyme.shallow(<Heading id="h1" level={1} text={"Heading 1"} />);
    expect(wrapper.find("h1")).to.have.lengthOf(1);
  });
  it("Can pass heading content as child", () => {
    wrapper = Enzyme.shallow(
      <Heading id="h1" level={1}>
        Text
      </Heading>
    );
    expect(wrapper.find("h1")).to.have.lengthOf(1);
  });

  // TODO: check that header children are links
  // it("Throws error when invalid heading is passed as child", () => {
  //   expect(() => Enzyme.mount(<Heading id="h1" level={3}><span>oh no</span></Heading>))
  //   .to.throw("Headings can only be plain text or bold");
  // });
  it("Can pass heading content as child span", () => {
    wrapper = Enzyme.shallow(
      <Heading id="h1" level={1}>
        <span>
          Text <b>hi</b>
        </span>
      </Heading>
    );
    expect(wrapper.find("h1")).to.have.lengthOf(1);
    expect(wrapper.find("span")).to.have.lengthOf(1);
  });
  it("uses child when both child and text are passed", () => {
    wrapper = Enzyme.shallow(
      <Heading id="h1" level={1} text={"Heading Text"}>
        Text
      </Heading>
    );
    expect(wrapper.find("h1")).to.have.lengthOf(1);
    expect(wrapper.find("h1").text()).to.equal("Text");
  });
  it("Has <a> tag when passed URL", () => {
    wrapper = Enzyme.shallow(
      <Heading id="h1" level={1} url="fake-url" text={"Heading 1"} />
    );

    expect(wrapper.find("h1")).to.have.lengthOf(1);
    expect(wrapper.find("a")).to.have.lengthOf(1);
  });
  it("<a> has class when passed urlClass", () => {
    wrapper = Enzyme.shallow(
      <Heading
        id="h1"
        level={1}
        url="fake-url"
        urlClass="fake-class"
        text={"Heading 1"}
      />
    );
    expect(wrapper.find("h1")).to.have.lengthOf(1);
    expect(wrapper.find("a").hasClass("fake-class")).to.equal(true);
  });
  it("Throws error when invalid heading number passed", () => {
    expect(() =>
      Enzyme.mount(<Heading id="h1" level={9} text={"Heading 9"} />)
    ).to.throw("Heading only supports levels 1-6");
  });
  it("Throws error when neither child nor text is passed", () => {
    expect(() => Enzyme.mount(<Heading id="h1" level={9} />)).to.throw(
      "Heading only supports levels 1-6"
    );
  });
  it("Throws error when heading with many children is passed", () => {
    expect(() =>
      Enzyme.shallow(
        <Heading id="h1" level={4}>
          <span>too</span>
          <span>many</span>
        </Heading>
      )
    ).to.throw("Please only pass one child into Heading, got span, span");
  });
  it("Uses custom display size", () => {
    wrapper = Enzyme.shallow(
      <Heading
        id="h1"
        level={1}
        text={"Heading with Secondary displaySize"}
        displaySize={DisplaySizes.Secondary}
      />
    );
    expect(wrapper.find("h1")).to.have.lengthOf(1);
    expect(wrapper.find("h1").hasClass("heading--secondary")).to.equal(true);
  });
});
