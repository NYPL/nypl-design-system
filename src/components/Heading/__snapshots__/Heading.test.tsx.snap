// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading renders the UI snapshot correctly 1`] = `
<h1
  className="chakra-heading css-1xdhyk6"
  id="basic"
>
  Heading text
</h1>
`;

exports[`Heading renders the UI snapshot correctly 2`] = `
<h1
  className="chakra-heading css-1xdhyk6"
  id="basicWithChildText"
>
  Heading text
</h1>
`;

exports[`Heading renders the UI snapshot correctly 3`] = `
<h1
  className="chakra-heading css-1xdhyk6"
  id="customDisplaySize"
>
  Heading with Secondary size
</h1>
`;

exports[`Heading renders the UI snapshot correctly 4`] = `
<h6
  className="chakra-heading css-1xdhyk6"
  id="otherLevel"
>
  Heading level six
</h6>
`;

exports[`Heading renders the UI snapshot correctly 5`] = `
<h1
  className="chakra-heading css-1xdhyk6"
  id="withLink"
>
  <a
    className="css-1xdhyk6"
    href="fake-url"
    id="withLink-link"
    rel={null}
    target={null}
  >
    Heading text is a link
  </a>
</h1>
`;

exports[`Heading renders the UI snapshot correctly 6`] = `
<h1
  className="chakra-heading css-1xdhyk6"
  id="withCustomLink"
>
  Part of the heading text is
  <a
    className="custom-link"
    href="fake-url"
  >
    <span>
      a link
    </span>
  </a>
</h1>
`;

exports[`Heading renders the UI snapshot correctly 7`] = `
<h2
  className="chakra-heading css-10g9ftz"
  id="chakra"
>
  Heading
</h2>
`;

exports[`Heading renders the UI snapshot correctly 8`] = `
<h2
  className="chakra-heading css-1xdhyk6"
  data-testid="props"
  id="props"
>
  Heading
</h2>
`;
